test/e2e/apimachinery/aggregator.go

```diff
@@ -536,6 +543,132 @@ func TestSampleAPIServer(f *framework.Framework, aggrclient *aggregatorclient.Cl
 	}
 	framework.ExpectEqual(locatedWardle, true, "Unable to find v1alpha1.wardle.example.com in APIServiceList")
 
+	// As the APIService doesn't have any labels currently set we need to
+	// set one so that a watch can track events based on the following label.
+	ginkgo.By("patch the APIService")
+	apiServiceName := "v1alpha1.wardle.example.com"
+	apiServiceClient := aggrclient.ApiregistrationV1().APIServices()
+	apiServicePatch, err := json.Marshal(map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"labels": map[string]string{"apiservice": "patched"},
+		},
+	})
+	framework.ExpectNoError(err, "failed to Marshal APIService JSON patch")
+	patchedResult, err := apiServiceClient.Patch(context.TODO(), apiServiceName, types.StrategicMergePatchType, []byte(apiServicePatch), metav1.PatchOptions{})
+	framework.ExpectNoError(err, "failed to patch APIService")
+	framework.Logf("APIService labels: %v", patchedResult.Labels)
+
+	patchedApiService, err := apiServiceClient.Get(context.TODO(), apiServiceName, metav1.GetOptions{})
+	framework.ExpectNoError(err, "Unable to retrieve api service %s", apiServiceName)
+	framework.Logf("APIService labels: %v", patchedApiService.Labels)
+
+	w := &cache.ListWatch{
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			options.LabelSelector = "apiservice=patched"
+			return apiServiceClient.Watch(context.TODO(), options)
+		},
+	}
+
+	apiServiceList, err := apiServiceClient.List(context.TODO(), metav1.ListOptions{LabelSelector: "apiservice=patched"})
+	framework.ExpectNoError(err, "failed to list API Services")
+
+	ginkgo.By("updating the APIService Status")
+	var statusToUpdate, updatedStatus *apiregistrationv1.APIService
+
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		statusToUpdate, err = apiServiceClient.Get(context.TODO(), apiServiceName, metav1.GetOptions{})
+		framework.ExpectNoError(err, "Unable to retrieve api service %s", apiServiceName)
+
+		statusToUpdate.Status.Conditions = append(statusToUpdate.Status.Conditions, apiregistrationv1.APIServiceCondition{
+			Type:    "StatusUpdate",
+			Status:  "True",
+			Reason:  "E2E",
+			Message: "Set from e2e test",
+		})
+
+		updatedStatus, err = apiServiceClient.UpdateStatus(context.TODO(), statusToUpdate, metav1.UpdateOptions{})
+		return err
+	})
+	framework.ExpectNoError(err, "Failed to update status. %v", err)
+	framework.Logf("updatedStatus.Conditions: %#v", updatedStatus.Status.Conditions)
+
+	ginkgo.By("watching for the APIService to be updated")
+	ctx, cancel := context.WithTimeout(context.Background(), apiServiceRetryTimeout)
+	defer cancel()
+	_, err = watchtools.Until(ctx, apiServiceList.ResourceVersion, w, func(event watch.Event) (bool, error) {
```

Comment:
this `watchtools.Until` looks almost identical to the one at line 648. TIOLI: you can DRY them up with a closure that returns a `watchtools.ConditionFunc`. E.g.,

```go
condFunc := func(condType apiregistrationv1.APIServiceConditionType, condReason, condMsg string) watchtools.ConditionFunc {
     return func(event watch.Event) (bool, error) {
     // ....
     }
}     
_, err = watchtools.Until(ctx, apiServiceList.ResourceVersion, w, condFunc("StatusUpdate", "E2E", "Set from e2e test"))
// ...
_, err = watchtools.Until(ctx, apiServiceList.ResourceVersion, w, condFunc("StatusPatched", "", ""))     
```